File: backend/src/indexing/indexer.ts
import type { PoolClient } from "pg";
import { SQL_UPDATE_DOCUMENT_SEARCH_VECTOR } from "../sql/indexer";

/**
 * Indexer handles final indexing for search backends.
 * Responsibilities:
 * - Update full-text search vectors (GIN indexes)
 * - Manage search metadata
 * - Ensure searchability across organization boundaries
 */

export async function updateSearchIndex(
  client: PoolClient,
  documentId: string
): Promise<void> {
  // Update GIN search vector for full-text search
  await client.query(SQL_UPDATE_DOCUMENT_SEARCH_VECTOR, [documentId]);
}

/**
 * Orchestrate the full ingestion pipeline:
 * normalize → chunk → index → resolve permissions
 */
export async function finalizeSearchability(
  client: PoolClient,
  documentId: string
): Promise<void> {
  // Ensure document is fully indexed for search queries
  await updateSearchIndex(client, documentId);
}

File: apps/api/app/indexing/indexer.py
from __future__ import annotations

import app.sql.indexer as sql_indexer


async def update_search_index(conn, document_id: str) -> None:
    await conn.execute(sql_indexer.update_document_search_vector(document_id))


async def finalize_searchability(conn, document_id: str) -> None:
    await update_search_index(conn, document_id)
